### PR TITLE
ArangoDB:  statistics reader for ArangoDB

### PR DESCRIPTION
**Pull request by matthewvon**
ChangeLog: ArangoDB plugin: new plugin for statistics reader

This is the initial implementation of a statistics reader for ArangoDB.

The basic code was copied from other contributors, so left those copyrights in place.

Requesting all comments about what is needed to get this into collectd permanently.

Yes, I am planning an arangodb_test.c file.  Just not there today.  Still proofing statistics.

### CODE REVIEW

**Comment by mkobyli** [comment]
Could you please address the check issues, like changelog, clang-format and build fail?

**Comment by dago** [comment]
@matthewvon Would you be willing to rebase this PR and resolve the conflicts in the code?

**Comment by matthewvon** [comment]
Honestly thought this was going to die.  Would be happy to rebase, likely tomorrow, in hopes of integration to main product.

**Comment by matthewvon** [comment]
Finally working issues.

**Comment by dago** [comment]
@matthewvon Would you mind adding the docs and configuration examples for this plugin as reported from the "New Plugin" CI check?

**Comment by matthewvon** [comment]
@dago  I see that as part of the checks list (or at least implied to me).  So answer is "yes".

File: configure.ac [hunk 1]
```diff
@@ -6990,6 +6990,7 @@ AC_PLUGIN([apache],              [$with_libcurl],             [Apache httpd stat
 AC_PLUGIN([apcups],              [yes],                       [Statistics of UPSes by APC])
 AC_PLUGIN([apple_sensors],       [$with_libiokit],            [Apple hardware sensors])
 AC_PLUGIN([aquaero],             [$with_libaquaero5],         [Aquaero hardware sensors])
+AC_PLUGIN([arangodb],            [$with_libcurl],             [ArangoDB statistics])
```
Comment (rpv-tomsk) [inline configure.ac]: Also add the appropriate line in the area of line L7430.

File: Makefile.am [hunk 2]
```diff
@@ -724,6 +724,16 @@ aquaero_la_LDFLAGS = $(PLUGIN_LDFLAGS) $(BUILD_WITH_LIBAQUAERO5_LDFLAGS)
...
 endif
 
+if BUILD_PLUGIN_ARANGODB
+pkglib_LTLIBRARIES += arangodb.la
+arangodb_la_SOURCES = src/arangodb.c
+arangodb_la_CFLAGS = $(AM_CFLAGS) $(BUILD_WITH_LIBCURL_CFLAGS)
+arangodb_la_CPPFLAGS = $(AM_CPPFLAGS) $(BUILD_WITH_LIBYAJL_CPPFLAGS)
+arangodb_la_LDFLAGS = $(PLUGIN_LDFLAGS) $(BUILD_WITH_LIBYAJL_LDFLAGS)
+arangodb_la_LIBADD = $(BUILD_WITH_LIBCURL_LIBS) $(BUILD_WITH_LIBYAJL_LIBS)
+
```
Comment (rpv-tomsk) [inline Makefile.am]: Please remove this empty line.

File: src/arangodb.c [hunk 3]
```diff
@@ -0,0 +1,1201 @@
...
+  
+} // cluster_verify_options
+
+
+//
+// collectd callback every "interval" to poll cluster,
+//  used to initiate poll to get cluster configuration
+//
+static int cluster_read(user_data_t *ud) {
+  int ret_val;
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Please declare variables on their first use.
https://github.com/collectd/collectd/blob/master/docs/review_comments.md#define-variables-on-first-use

File: src/arangodb.c [hunk 4]
```diff
@@ -0,0 +1,1201 @@
...
+static bool service_data_set_done = false;
+
+static pthread_mutex_t rocks_data_set_mutex = PTHREAD_MUTEX_INITIALIZER;
+static bool rocks_data_set_done = false;
+
+
+//
+// entry point from collectd daemon
+//
+void module_register(void) {
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Such entry points typically placed at end of file, in Collectd and in other project.
Please move it to the end.
Thanks.

File: src/arangodb.c [hunk 5]
```diff
@@ -0,0 +1,1201 @@
...
+// collectd callback every "interval" to poll cluster,
+//  used to initiate poll to get cluster configuration
+//
+static int cluster_read(user_data_t *ud) {
+  int ret_val;
+  cluster_t * cluster;
+  
+  ret_val = 0;
+
+  if ( NULL != ud && NULL != ud->data) {
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Use `fast return`  in your code. Do not put constant in front of variable in conditionals.
Reorder such (this and all similar) code to:
```
if ((ud == NULL) || (ud->data == NULL)) {
  ERROR("....");
  return -1;
}

... code ..

```

File: src/arangodb.c [hunk 6]
```diff
@@ -0,0 +1,1201 @@
...
+                                         });
+
+        
+      } else {
+        // cluster of servers
+        yajl_val health_node = NULL;
+
+        ret_val = curl_perform(&cluster->curlinfo, &cluster->endpoint, "/_admin/cluster/health", &health_node);
+
+        if (0 == ret_val) {
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Do not put constant in front of variable in conditionals.

File: src/arangodb.c [hunk 7]
```diff
@@ -0,0 +1,1201 @@
...
+      WARNING("arangodb plugin: response JSON did not parse.");
+    } // else
+  } // if
+
+  return ret_val;
+
+} // curl_perform
+
+
+static size_t cluster_write_callback(void *buf,
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Can you please rename this function into `cluster_curl_callback` to don't confuse.
You don't write nothing into cluster.

File: src/arangodb.c [hunk 8]
```diff
@@ -0,0 +1,1201 @@
...
+  curl_global_init(CURL_GLOBAL_ALL);
+
+  return 0;
+} // arangodb_init
+
+
+//
+// Config level one:  cluster / singleserver key word
+//
+static int arangodb_complex_config(oconfig_item_t * config_item) {
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Typically, order of functions in code is upwards by call order.
Then you do not need to do forward  declarations.

So, preferred order looks like

... and so on.
arangodb_add_cluster
arangodb_complex_config
arangodb_init
module_register

File: src/arangodb.c [hunk 9]
```diff
@@ -0,0 +1,1201 @@
...
+  int ret_val;
+  cluster_t * cluster;
+  
+  ret_val = 0;
+
+  if ( NULL != ud && NULL != ud->data) {
+    char * role = NULL;
+    cluster = (cluster_t *)ud->data;
+
+    role = arangodb_get_role(&cluster->curlinfo, &cluster->endpoint);
```
Comment (rpv-tomsk) [inline src/arangodb.c]: In your implementation, if call to `arangodb_get_role` fails, code mades the wrong decisions, I think.

File: src/arangodb.c [hunk 10]
```diff
@@ -0,0 +1,1201 @@
...
+
+        ret_val = curl_perform(&cluster->curlinfo, &cluster->endpoint, "/_admin/cluster/health", &health_node);
+
+        if (0 == ret_val) {
+          cluster_decode_health(health_node);
+        } // if
+
+        yajl_tree_free(health_node);
+
+      } // else
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Such comments are not useful.

File: src/arangodb.c [hunk 11]
```diff
@@ -0,0 +1,1201 @@
...
+        value_t value;
+        value_list_t vl = VALUE_LIST_INIT;
+        bool skip = false;
+        int ret_val;
+        
+        vl.values = &value;
+        vl.values_len = 1;
+        vl.time = now;
+        // default interval?
+        ssnprintf(vl.host, DATA_MAX_NAME_LEN, "%s", service->endpoint.host);
```
Comment (rpv-tomsk) [inline src/arangodb.c]: Don't use DATA_MAX_NAME_LEN.
Please use `sizeof()` (sizeof(vl.host) and so on).

**Comment by matthewvon** [comment]
I am not going to resolve the code changes requested.  I am therefore closing this pull request.

**Pull request closed without merging**